### PR TITLE
Improve fit

### PR DESCRIPTION
**Pull request by hwpang**
This builds on top of https://github.com/zadorlab/PySIDT/pull/7 and https://github.com/zadorlab/PySIDT/pull/7 should be merged in first. 

I changed the multieval tree fitting scheme. I found that the current fitting scheme does not guarantee smaller values for nodes closer to the terminal nodes. Instead, I took inspiration from the gradient boosting to do the fitting layer by layer, from top of the tree (root) to the bottom (terminal nodes).

### CODE REVIEW

**Comment by mjohnson541** [comment]
This approach looks very sound to me, but can we do some rough accuracy benchmarks before proceeding?

**Comment by hwpang** [comment]
Did some preliminary benchmarking on this change. Using 100 data points as hold out test set, 100 as validation set, and 800 as training set. Using 50 nodes.

# Fitting all together (current main)
all_together

# Layered fitting (this PR)
layered_tree

**Comment by mjohnson541** [comment]
Looks good! I take it this is J/mol E0 or H298? Is this ready for review?

**Comment by hwpang** [comment]
Yeah this is E0. I need to double check the unit. Yeah this is ready for review.

File: pysidt/sidt.py [hunk 1]
```diff
@@ -636,47 +637,55 @@ def fit_tree(self, data=None, check_data=True, alpha=0.1):
             self.setup_data(data, check_data=check_data)
             self.descend_training_from_top(only_specific_match=True)
 
-        # generate matrix
-        A = sp.csc_matrix((len(self.datums), len(self.nodes)))
+        max_depth = self.assign_depths()
```
Comment (mjohnson541) [inline pysidt/sidt.py]: Rather than reassigning the depths every iteration shouldn't we just handle this when we add new nodes (assigning parent.depth + 1)?

File: pysidt/sidt.py [hunk 2]
```diff
@@ -636,47 +637,55 @@ def fit_tree(self, data=None, check_data=True, alpha=0.1):
...
-            max_iter=1000000000,
-            selection="random",
-        )
+        for depth in range(max_depth + 1):
+            nodes = [node for node in self.nodes.values() if node.depth == depth]
 
-        pred = clf.fit(A, y)
-        self.data_delta = A * clf.coef_ - y
+            # generate matrix
+            A = sp.lil_matrix((len(self.datums), len(nodes)))
```
Comment (mjohnson541) [inline pysidt/sidt.py:704]: Why lil over any other type of sparse matrix?

File: pysidt/sidt.py [hunk 3]
```diff
@@ -636,47 +637,55 @@ def fit_tree(self, data=None, check_data=True, alpha=0.1):
...
-            self.node_uncertainties = {node.name: 1.0 for i, node in enumerate(nodes)}
 
-        for i, val in enumerate(clf.coef_):
-            nodes[i].rule = val
+            lasso = clf.fit(A, y)
+            preds = A * clf.coef_
+            self.data_delta = preds - y
+
+            if A.shape[1] != 1:
+                node_uncertainties = (
```
Comment (mjohnson541) [inline pysidt/sidt.py]: I don't think the analytical formula being used here is still valid after breaking up the optimization by depth this way. I think that this formula based approach is still good enough for estimating uncertainties for node splitting in this case, but I think we may need to construct the full A at the end of the optimizations to evaluate this.

File: pysidt/sidt.py [hunk 4]
```diff
@@ -27,9 +27,10 @@ def __init__(
         self.parent = parent
         self.children = children
         self.name = name
+        self.depth = 0
```
Comment (mjohnson541) [inline pysidt/sidt.py]: Should we add this as a input parameter for Node?

File: pysidt/decomposition.py [hunk 5]
```diff
@@ -8,6 +8,8 @@ def atom_decomposition(mol):
 
 
 def atom_decomposition_noH(mol):
+    clear_decomposition(mol)
```
Comment (mjohnson541) [inline pysidt/decomposition.py]: Why is this necessary? Why is someone feeding in data with labels that we want the algorithm to ignore?

I having a little trouble thinking of an example where I think it's smart to do this...but one could imagine using this kind of predictor to predict rate coefficients for a given family what we would typically use a single evaluation tree for (a la reaction GAV) in which case the input data would have labels that are important for prediction...

We can certainly add checks that will throw errors if things don't look in order...but I'm not sure why we can't expect the user to provide a dataset with sensible labeling...

File: pysidt/sidt.py [hunk 6]
```diff
@@ -339,6 +339,23 @@ def read_nodes(file):
 
 
 class MultiEvalSubgraphIsomorphicDecisionTree(SubgraphIsomorphicDecisionTree):
+    """
+    Makes prediction for a molecule based on multiple evaluations.
+
+    Args:
+        `decomposition`: method to decompose a molecule into groups.
```
Comment (mjohnson541) [inline pysidt/sidt.py]: Broadly this is true, we are decomposing them into group contributions but on a technical standpoint this doesn't make RMG Group objects so maybe "groups" => "substructure contributions"

File: pysidt/sidt.py [hunk 7]
```diff
@@ -339,6 +339,23 @@ def read_nodes(file):
...
+        `root_group`: root group for the tree
+        `nodes`: dictionary of nodes for the tree
+        `n_splits`: number of splits to try for each node. Default is 1.
```
Comment (mjohnson541) [inline pysidt/sidt.py]: n_splits is the minimum number of disconnected structures that should be in the group. For example in bimolecular families if you form a bond between the two groups doing H_Abstraction you only have one structure with both reactive sites and you're actually doing intra_H_migration instead of H_Abstraction. n_splits stops us from forming improper internal bonds.

File: pysidt/sidt.py [hunk 8]
```diff
@@ -339,6 +339,23 @@ def read_nodes(file):
...
+        `iter_max`: maximum number of iterations for each split. Default is 2.
```
Comment (mjohnson541) [inline pysidt/sidt.py]: inter_max: maximum number of times the extension generation algorithm is allowed to expand structures looking for additional splits
iter_item_cap: the maximum number of structures the extension generation algorithm can send for expansion

**Review by mjohnson541 — COMMENTED** [review]
Thanks! Mostly looks good!

File: pysidt/sidt.py (hunk 2 above)
Comment (hwpang) [reply pysidt/sidt.py:704]: Using the `sp.csc_matrix` threw a warning along the line of "changing the size of `csc_matrix` is expensive, use `lil_matrix` instead", so I took the suggestion.

File: pysidt/sidt.py (hunk 7 above)
Comment (hwpang) [reply pysidt/sidt.py]: hmmm this keyword variable name is really confusing. We should change it. What about `n_disconnected_struc`? "Splits" is a term with associated meaning in decision tree, so I don't think we should use this word like this.

**Comment by hwpang** [comment]
@mjohnson541 I've addressed most of your comments. I think this PR is ready for another review.

File: pysidt/sidt.py (hunk 7 above)
Comment (hwpang) [reply pysidt/sidt.py]: From subgroup meeting: change to `n_strucs`.

File: pysidt/sidt.py [hunk 9]
```diff
@@ -657,18 +657,29 @@ def generate_tree(
...
         """
         if data:
             self.setup_data(data, check_data=check_data)
             self.descend_training_from_top(only_specific_match=True)
 
-        max_depth = max([node.depth for node in self.nodes.values()])
+        if fit_rule:
+            self.fit_rule(alpha=alpha)
+        
+        if estimate_full_uncertainty:
```
Comment (mjohnson541) [inline pysidt/sidt.py]: This should be default for now and we should remove the layer-by-layer uncertainties. Perhaps we'll need a faster way later, but I don't think we trust the layer by layer uncertainties yet.

**Comment by hwpang** [comment]
@mjohnson541 This is ready for another review

File: pysidt/sidt.py [hunk 10]
```diff
@@ -53,7 +53,7 @@ def __init__(
         self,
         root_group=None,
         nodes=None,
-        n_splits=1,
```
Comment (mjohnson541) [inline pysidt/sidt.py:55]: n_splits is also in the extensions.py file.

File: pysidt/sidt.py [hunk 11]
```diff
@@ -346,7 +346,7 @@ class MultiEvalSubgraphIsomorphicDecisionTree(SubgraphIsomorphicDecisionTree):
         `decomposition`: method to decompose a molecule into substructure contributions.
         `root_group`: root group for the tree
         `nodes`: dictionary of nodes for the tree
-        `n_strucs`: number of splits to try for each node. Default is 1.
+        `n_strucs`: number of disconnected structures that should be in the group. Default is 1.
```
Comment (mjohnson541) [inline pysidt/sidt.py]: Hmm...so suppose you have an intra_H_migration we could conceivably start with an R. and separate a R-H that eventually become the same molecule...since we only create bonds in this process (never destroy bonds) it might be best to say this is the "minimum number of disconnected structures that can be in the group".

File: pysidt/sidt.py [hunk 12]
```diff
@@ -657,25 +657,19 @@ def generate_tree(
...
             self.setup_data(data, check_data=check_data)
             self.descend_training_from_top(only_specific_match=True)
 
-        if fit_rule:
-            self.fit_rule(alpha=alpha)
+        self.fit_rule(alpha=alpha)
         
-        if estimate_full_uncertainty:
-            self.estimate_full_uncertainty()
+        self.estimate_full_uncertainty()
```
Comment (mjohnson541) [inline pysidt/sidt.py]: Shouldn't this be just "estimate_uncertainty" now since there isn't a partial vs full anymore?

**Review by mjohnson541 — COMMENTED** [review]
Looks mostly good, just a few comments.

File: pysidt/sidt.py (hunk 11 above)
Comment (hwpang) [reply pysidt/sidt.py]: I change `n_strucs` to `n_strucs_min` to reflect that it's the minimum

**Comment by hwpang** [comment]
@mjohnson541 Thanks for the review and the comments! I've resolved everything, and it's ready for another review.

**Comment by mjohnson541** [comment]
Alright! Looks good! Can you squash?

**Comment by hwpang** [comment]
@mjohnson541 Rebased! Can you review and merge after the tests pass?
